### PR TITLE
TypeResolver optimizations

### PR DESCRIPTION
**Pull request by azabluda**
Fixes https://github.com/6bee/aqua-core/issues/15

Please check 88f9a3b commit message. There is still an open question before we can merge this.

### CODE REVIEW

**Comment by 6bee** [comment]
Had no actual use case yet for *non-anonymous* types with identical names and property names which differ in property types only. Hence, you may move forward and get rid of the corresponding tests. 

If needed, we can re-add this later and distinguish between *generic* and *non-generic* types within `TypeResolver`. 

One shortcoming I expect is the caching-key used in `TypeResolver` which threats a type`s name the same as its property names. Hence
```
public class Foo { public int Bar { get; set; } }
```
was probably the same as 
```
public class Bar { public int Foo { get; set; } }
```
Here you might want to create a specific type, used as caching-key, handling the type's name separate from its property names in terms of equality. 

While you’re at it, could you also remote the `ref` keyword in `TypeResolver.IsValid` param list? 

Otherwise I don’t see any big issued, but rather some complexity gone 👍

**Comment by azabluda** [comment]
> the caching-key used in TypeResolver which threats a type`s name the same as its property names

Since the class name is always on the first position in the array and is usually prefixed with namespace, I don't see how it can lead to lookup confusions. A specific type for the caching key is something we may eventually consider as an obvious solution if we ever run into problems with the current one. TBH I was rather thinking about a simple comma- (newline-, pipe-, whatever-) separated string as a key, so we could rely on well developed and highly optimized string lookup algorithms. Shall we do this change?

> remove the ref keyword in TypeResolver.IsValid param list?

Removed.

**Comment by azabluda** [comment]
I added two more commits to illustrate my idea. Before merging I would prefer squashing a few iterations to shape a cleaner history.

**Comment by 6bee** [comment]
Using a single concatenated string as cache-key is fine with me. Although I'm undecided whether the sort oder or the property names should be relevant. 
Feel free to squash if you prefer. I have no strong opinion on this as sometimes a more comprehensive history may help understanding as well.

**Comment by 6bee** [comment]
I'm happy to merge once you're ok ;-)

**Comment by azabluda** [comment]
Valid point! I was thinking about sorting of property names myself, because `TypeInfo.Properties` don't seem to have any particular order (especially when we are dealing `TypeInfo`s coming from different machines). I've included `OrderBy(name)` in my final proposal. Ready for merge.

**Comment by 6bee** [comment]
I'was just trying to think a bout a case where sorting property names might lead to failure are well, e.g. anonymous types with only difference in the order of their properties:
```
var t1 = new { X = 1, Y = 2.0 }; // <>anonymous1<int,double>
var t2 = new { Y = 2.0, X = 1 }; // <>anonymous2<double,int>
var t3 = new { Y = 2, X = 1.0 }; // <>anonymous2<int,double>
```

However, so far I couldn't find any issues since this should result in different types i.e. types with different names. Hence, the resulting cache-keys should look something like: `"<>anonymous1 X Y"` and `"<>anonymous2 X Y"`

I guess we're fine for the time being, but might actually not in future since we heavily rely on the compiler. Any thoughts on this?

**Comment by azabluda** [comment]
In your examples we will get different names only if both anonymous types are coming from the same assembly, which in reality isn't always the case. So your concerns are valid.

I think the problem is easier to address in [`TypeInfo` class](https://github.com/6bee/aqua-core/blob/4.0.0-beta-003/src/Aqua/TypeSystem/TypeInfo.cs#L79) by adding `OrderBy(p => p.MetadataToken)` according to https://stackoverflow.com/a/31253200. Then my `OrderBy(name)` needs to be taken out.

Unfortunately the `MetadataToken` trick is nowhere officially documented and may just stop working one day, but so far I have no better idea.

**Comment by 6bee** [comment]
It's hard to tell whether that would actually help when dealing with different assemblies. Since [msdn](https://msdn.microsoft.com/en-us/library/system.reflection.memberinfo.metadatatoken.aspx)  states `MetadataToken` is defined per Module. 
However, I think it's still better than sorting alphabetically.

**Comment by azabluda** [comment]
I pushed my proposal addressing the issue. Unfortunately we cannot use it in .NET Standard 1.3 due to lack of  `MetadataToken`.

File: src/Aqua/TypeSystem/TypeResolver.cs
Comment (6bee) [inline src/Aqua/TypeSystem/TypeResolver.cs]: `CollectionEquals` should definitely lead to an error in tests `When_resolving_different_anonymous_types_with_same_name_and_different_order_of_properties` whereas `SequenceEqual` would probably still leave a change for success.

Unfortunately, we don't run any unit test on `netstandard 1.3` since I don't know how to achieve this :-(

Comment (azabluda) [reply src/Aqua/TypeSystem/TypeResolver.cs]: Right, I simplified this part and rewrote my last commit all together.

**Review by 6bee — APPROVED** [review]
Looks great to me, thank you!

**Comment by 6bee** [comment]
resolved issue #15

**Pull request merged**
